### PR TITLE
feat: preconfigured personas from a folder of markdown files (closes #880)

### LINKED ISSUE #880: make a folder for default system prompts
make a folder that containers .md files. 
the sections of the .md file define the name, access group, and text of the prompt
at startup load these as pre-configured system prompts that a user could select. 
So like I might want different personas that are preloaded into ATLAS, .. with needing a mcp prompt and without needing to create my own. 
This allows the admin to easily create many new personas or system prompts by just changing .md files in a folder.

### PR DESCRIPTION
**Pull request by garland3**
Closes #880.

## What

Admins can define selectable system prompts ("personas") by dropping `.md` files in a folder. No code changes, no MCP prompt server, and no per-user prompt authoring needed.

```markdown
---
name: Research Assistant          # optional; defaults to the filename
description: Careful, source-first answers
access_group: research-team       # optional; one group or a list
order: 10                         # optional sort hint
---
You are a meticulous research assistant...
```

Search order (first folder containing any `.md` wins, so your folder fully replaces the packaged samples): `PERSONAS_DIR` → `<PROMPT_BASE_PATH>/personas/` → `prompts/personas/` → packaged `atlas/config/prompts/personas/`.

## Design answers to the questions on the issue

- **Format:** YAML frontmatter + body, matching the existing `.md` prompt templates and every other config asset in the repo.
- **No access group = visible to everyone.** Hiding by default would break the "just add a file" workflow the issue asks for. Gated personas go through the same `is_user_in_group` check as tools and models, and are hidden if that check raises (fail closed).

## How it works

- `atlas/modules/prompts/persona_library.py` parses and caches the folder; read once during startup so parse problems land in the startup log, not a user's first request.
- `GET /api/personas` lists what the caller may see; `GET /api/personas/{id}` returns one (404 for both missing and not-allowed, so the endpoint does not confirm gated personas exist).
- Frontend: `usePersonas` + a **Personas** section in the prompt picker. Selecting one sends its text as the turn's `custom_system_prompt`, exactly like a user-authored prompt (#153) — no server-side chat plumbing changed.
- `README.md`/`index.md`, dotfiles and `_`-prefixed files are skipped so a folder can document itself.

Docs: `docs/admin/personas.md`.

## Verification

- `bash test/atlas_tests.sh` — 2742 passed, 17 skipped (16 new persona tests).
- `npx vitest run` — 710 passed (7 new persona tests).
- `npm run lint` — 0 errors (4 pre-existing warnings).
- Live server: `/api/personas` returned the packaged samples; with an override folder containing an ungated, a `users`-gated and a `nobody-group`-gated persona, the list contained only the first two and `GET /api/personas/secret` returned 404.
- Browser: the **Personas** section renders, selecting one relabels the picker, and the resulting LLM call was 66 chars of messages instead of ~5.5k — i.e. the persona replaced the default system prompt end to end.

### CODE REVIEW

File: frontend/src/contexts/ChatContext.jsx [hunk 1]
```diff
@@ -601,13 +606,21 @@ export const ChatProvider = ({ children }) => {
...
+
 		const sent = sendMessage({
 			type: 'chat',
 			content,
 			model: currentModel,
 			selected_tools: toolsToSend,
-			selected_prompts: activeKeyIsUserPrompt ? [] : activePrompts,
-			custom_system_prompt: activeUserPrompt ? activeUserPrompt.content : undefined,
+			selected_prompts: (activeKeyIsUserPrompt || activeKeyIsPersona) ? [] : activePrompts,
+			custom_system_prompt: customSystemPrompt,
```
[the remark this answers is not in the document]
Comment (garland3) [reply frontend/src/contexts/ChatContext.jsx]: Fixed in 2bd1804 — and thank you, this was a real gap. Rather than loosening the gate I made the persona text never travel from the client: the frontend now sends only `persona_id`, and the websocket handler resolves the text through `resolve_persona_prompt()`, which re-checks the persona's access group for the authenticated user. So personas are independent of `custom_prompts_effective` (they need no per-user storage), while a hand-crafted client can neither select a persona it was never offered nor substitute text of its own.

Verified live with `FEATURE_CUSTOM_PROMPTS_ENABLED=false FEATURE_CHAT_HISTORY_ENABLED=false`: selecting a "Pirate" persona produced "Ahoy, matey!" and a 108-char message payload instead of the ~5.5k default system prompt.

File: atlas/modules/prompts/persona_library.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+            candidates.append(project_root / configured_path)
+
+    prompt_base = Path(settings.prompt_base_path)
+    if prompt_base.is_absolute():
+        candidates.append(prompt_base / PERSONA_DIR_NAME)
+    else:
+        candidates.append(project_root / prompt_base / PERSONA_DIR_NAME)
+
+    candidates.append(project_root / "prompts" / PERSONA_DIR_NAME)
+    candidates.append(atlas_root / "config" / "prompts" / PERSONA_DIR_NAME)
```
[the remark this answers is not in the document]
Comment (garland3) [reply atlas/modules/prompts/persona_library.py]: Fixed in 2bd1804 — correct, setuptools package-data globs do not recurse. Added `config/prompts/personas/*.md`, and verified by building the wheel: it now contains `atlas/config/prompts/personas/{README,code-reviewer,plain-language-editor,research-assistant}.md`.

File: frontend/src/components/PromptSelector.jsx [hunk 3]
```diff
@@ -178,6 +184,47 @@ const PromptSelector = () => {
...
+                    <div className="flex items-center justify-between gap-2">
+                      <div className="flex-1 min-w-0">
+                        <div className="font-medium text-gray-200 flex items-center gap-2">
+                          {isActive && <span className="text-blue-400">✓</span>}
+                          <span className="truncate">{p.name}</span>
+                          {isActive && <span className="text-xs text-blue-400">(active)</span>}
+                        </div>
+                        <div className="text-xs text-gray-400 mt-1 line-clamp-2">
+                          {p.description || p.content}
+                        </div>
```
[the remark this answers is not in the document]
Comment (garland3) [reply frontend/src/components/PromptSelector.jsx:251]: Fixed — the fallback now renders at most 160 characters (`personaPreview`), with a test covering a 5k-char persona.

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## [Unreleased]
 
+### PR #PRNUM - 2026-09-02
+- **Preconfigured personas (default system prompts) from a folder of markdown files** (closes #880): admins can drop `.md` files with YAML frontmatter (`name`, `description`, optional `access_group`, `order`) into a personas folder and Atlas offers each one as a selectable system prompt in a new **Personas** section of the prompt picker -- no code changes, no MCP prompt server, and no per-user prompt authoring. Personas with no `access_group` are visible to everyone; gated ones are filtered through the same authorization service as tools and models (and hidden if that check errors). Search order: `PERSONAS_DIR`, `<PROMPT_BASE_PATH>/personas/`, `prompts/personas/`, then the packaged samples in `atlas/config/prompts/personas/`. New endpoints `GET /api/personas` and `GET /api/personas/{id}`; see `docs/admin/personas.md`.
```
[the remark this answers is not in the document]
Comment (garland3) [reply CHANGELOG.md]: The `PRNUM` placeholder was a stopgap until the PR number existed; it became `### PR #882 - 2026-09-02` in 4580d6f, one commit after the one you reviewed. The entry is a single bullet, in line with recent entries such as PR #875.

File: atlas/routes/persona_routes.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+from fastapi import APIRouter, Depends, HTTPException
+
+from atlas.core.auth import is_user_in_group
+from atlas.core.log_sanitizer import get_current_user
+from atlas.modules.prompts.persona_library import get_persona_library
+
+logger = logging.getLogger(__name__)
+
+router = APIRouter(prefix="/api/personas", tags=["personas"])
```
[the remark this answers is not in the document]
Comment (garland3) [reply atlas/routes/persona_routes.py]: Fixed — the module no longer logs, so the logger and the `logging` import are gone.

**Comment by garland3** [comment]
## Summary
Adds admin-authored "personas": markdown files with YAML frontmatter loaded from a configurable folder at startup, exposed via `/api/personas`, selectable in the prompt picker, and resolved server-side from a `persona_id` on each chat turn with a per-request access-group re-check.

## Verdict
Request changes.

## Required changes
1. Fail closed when the frontmatter fails to parse at `atlas/modules/prompts/persona_library.py:170` — a `yaml.YAMLError` currently yields `access_groups == []`, which means "visible to everyone", so a typo in a gated persona's frontmatter publishes it to all users. Skip the file (or assign an unsatisfiable group) and update `test_broken_frontmatter_keeps_the_prompt` and `docs/admin/personas.md` accordingly.
2. Send `persona_id` off the active key rather than the fetched list at `frontend/src/contexts/ChatContext.jsx:626` — after a reload, or when `/api/personas` errors, the picker still shows the persona name but the turn goes out with no `persona_id` and runs on the default prompt.
3. Serialize the list endpoint with `to_dict(include_content=False)` at `atlas/routes/persona_routes.py:22` (add a short server-computed preview if the picker needs one) — it currently ships up to 100k chars of prompt body per persona on every page load for a 160-char subtitle.
4. Resolve group membership once per request in `personas_for_user` at `atlas/modules/prompts/persona_library.py:220`, and look the persona up by id before authorizing in `get_persona`/`resolve_persona_prompt` — today every gated persona costs a serial auth round-trip on each chat turn's hot path.
5. Add a test for the websocket persona branch at `atlas/main.py:724` covering: inline `custom_system_prompt` wins over `persona_id`; a persona still resolves when `custom_prompts_effective` is false; an unauthorized `persona_id` leaves `custom_system_prompt` as `None`.
6. Surface persona load failure in the UI at `frontend/src/hooks/usePersonas.js:31` — consume the already-exported `personasError` to show the section with an inline message and a retry, instead of rendering nothing (indistinguishable from "none configured").
7. Correct the stale docstrings at `frontend/src/hooks/usePersonas.js:8`, `frontend/src/hooks/chat/useSelections.js:19`, and `atlas/routes/persona_routes.py:4` — all three still say the client sends the persona text as `custom_system_prompt`, contradicting the server-side resolution the security model depends on.

## Evidence
Evidence: 7/10 — the backend is well demonstrated (23 new automated tests, a live end-to-end validation script, measured prompt-substitution output), but the user-facing picker section is asserted only in prose. The most useful missing artifact is a screenshot of the prompt selector with the Personas section populated and the trigger relabelled after selection.

## Notes
- Close the dropdown on selection: `handlePromptSelect` omits `setIsOpen(false)` unlike the Default Prompt row — `frontend/src/components/PromptSelector.jsx:205`.
- `access_groups` is returned to every caller; a boolean `restricted` would avoid leaking internal group names — `atlas/routes/persona_routes.py:22`.
- Duplicate-id shadowing within a directory logs at info level and is untested; a warning plus a test would catch `code-reviewer.md` vs `code_reviewer.md` — `atlas/modules/prompts/persona_library.py:186`.

---
_Posted automatically by **AGENT-REVIEW-BOT-3**_

**Comment by garland3** [comment]
Addressed the review in 62f8570 (plus 374f107 for the folder default and workspace capture):

1. **Fail closed on frontmatter parse** -- invalid YAML, or frontmatter that is not a mapping, now skips the file with an error log, so a typo in a gated persona's access_group cannot publish it to everyone. `test_broken_frontmatter_skips_the_file` and `test_non_mapping_frontmatter_skips_the_file` replace the old keeps-the-prompt test; docs updated.
2. **persona_id from the active key** -- the send path now derives the id with `personaIdFromKey(activeKey)` instead of looking it up in the fetched list, so a reload or a failed /api/personas no longer silently runs the default prompt. A genuinely gone or gated-away persona still resolves to None server-side (fail closed on the side that owns the folder).
3. **List endpoint without content** -- `GET /api/personas` returns `to_dict(include_content=False)` with a 160-char server-computed `preview`; the picker prefers it, and chat export falls back to it. `GET /api/personas/{id}` keeps the full content.
4. **Group checks once per request** -- `personas_for_user` now checks each distinct group exactly once, concurrently via asyncio.gather, and reuses the results; the new `persona_for_user` looks up by id *before* authorizing and backs both the get-by-id route and `resolve_persona_prompt`, so one persona costs at most its own group checks. Covered by `test_each_distinct_group_is_checked_once` and `test_persona_for_user_authorizes_only_that_persona`.
5. **Websocket branch tests** -- the branch is now `resolve_chat_system_prompt()` with direct tests: inline custom_system_prompt wins over persona_id; a persona resolves when custom_prompts_effective is false (and an inline prompt sent while the flag is off is ignored); an unauthorized persona_id leaves the prompt as None.
6. **Load failure surfaced** -- the picker consumes `personasError` and renders the Personas section with an inline message and a Retry button (`fetchPersonas`) instead of rendering nothing.
7. **Docstrings corrected** in usePersonas.js, useSelections.js, and persona_routes.py -- all now describe server-side resolution from persona_id.

Also in this round, per request: the default personas folder is now `config/personas/` (`<APP_CONFIG_DIR>/personas/`), and the active persona is captured in workspaces like any other active prompt -- the workspace summary names it, and persona keys stay out of the MCP selected_prompts set.

Deferred from the notes: dropdown close-on-select (shared handler for MCP/user prompts -- behavior change beyond personas), the access_groups payload field, and the duplicate-id log level. Happy to take those as a follow-up.

**Comment by garland3** [comment]
## Summary
Adds a folder-based persona library: markdown personas with YAML frontmatter and optional group gating, a `GET /api/personas` REST surface, server-side prompt resolution from `persona_id` on the chat websocket path, and a Personas section in the prompt picker.

## Verdict
Block — critical issues found.

## Required changes
1. Make unparseable frontmatter fail closed at `atlas/modules/prompts/persona_library.py:170` — on `yaml.YAMLError` the parser continues with `metadata = {}`, so `access_groups` becomes `[]` and a typo in a gated persona's frontmatter publishes it to every user. Return `None` (log the filename at error level) or assign an unsatisfiable sentinel group; update `test_broken_frontmatter_keeps_the_prompt` and the matching line in `docs/admin/personas.md`.
2. Add `test_broken_frontmatter_on_a_gated_persona_is_not_public` in `atlas/tests/test_persona_library.py:59` asserting a file whose raw text contains `access_group` never yields `access_groups == []` after a parse failure — the current test locks in the fail-open behavior.
3. Resolve `access_group` / `access_groups` explicitly at `atlas/modules/prompts/persona_library.py:180` — a present-but-null `access_group:` key suppresses the `access_groups` fallback and makes the persona public.
4. Send `persona_id` off the persisted key at `frontend/src/contexts/ChatContext.jsx:626` (`activeKeyIsPersona ? personaIdFromKey(activeKey) : undefined`) — deriving it from the fetched list means a pending or failed `/api/personas` sends the turn with no persona while the picker still shows the persona name.
5. Add a test for the persona branch at `atlas/main.py:724` covering the three contracts it encodes: inline `custom_system_prompt` wins over `persona_id`, a persona resolves when `custom_prompts_effective` is false, and an unauthorized id leaves `custom_system_prompt` as `None`. Extract a `resolve_turn_system_prompt(data, user_email, flag_enabled)` helper if the handler is hard to drive.
6. Look the persona up by id before authorizing in `resolve_persona_prompt` / `get_persona` (`atlas/modules/prompts/persona_library.py:220`) — every chat turn currently authorizes the entire library serially, so latency scales with the number of gated personas.
7. Serialize the list endpoint with `to_dict(include_content=False)` plus a server-computed `preview` field at `atlas/routes/persona_routes.py:22`, and drop `access_groups` from the payload — page load ships up to 100k chars of prompt body per persona plus internal authorization group names that no client reads.

## Evidence
Evidence: 7/10. The backend is well covered (16 backend + 7 frontend tests, a committed end-to-end validation script, a measured payload-size check of the substitution), but the user-facing half is asserted only in prose; the most important missing artifact is a screenshot of the prompt picker with the Personas section populated and the trigger relabelled after selection.

## Notes
- Duplicate-id shadowing (`code-reviewer.md` vs `code_reviewer.md`) is resolved by byte order and logged at info; raise to warning with both paths and test it — `atlas/modules/prompts/persona_library.py:186`.
- `personasError` is exported but never rendered, so a 500 looks identical to "no personas configured" — `frontend/src/components/PromptSelector.jsx:196`.
- `handlePromptSelect` omits `setIsOpen(false)`, unlike the adjacent Default Prompt row, so the menu stays open after selecting — `frontend/src/components/PromptSelector.jsx:205`.

---
_Posted automatically by **AGENT-REVIEW-BOT-3**_

**Comment by garland3** [comment]
## Summary
Adds a filesystem-backed persona library (loader module, two REST endpoints, chat-turn system-prompt resolution) plus a new Personas section in the frontend prompt selector.

## Verdict
Request changes.

## Required changes
1. Resolve the two group keys explicitly at `atlas/modules/prompts/persona_library.py:180` — `metadata.get("access_group", metadata.get("access_groups"))` only falls back when the key is *absent*, so a bare `access_group:` line yields `None` → `[]`, which means "visible to every authenticated user" and publishes a gated persona. Prefer failing closed (skip the file with an error log) when a group key is present but normalizes empty.
2. Add a test asserting that frontmatter containing an `access_group` key never yields `access_groups == []` — in `atlas/tests/test_persona_library.py`, covering both the null-value and empty-list forms.
3. Reject non-list/str group values in `_normalize_groups` at `persona_library.py:64` — a mapping is currently accepted and stringified into a junk group name rather than erroring.
4. Treat an empty-string inline `custom_system_prompt` as unset in `resolve_chat_system_prompt` (`persona_library.py:392`) — today it wins precedence and silently discards the user's selected persona.

## Evidence
Evidence: 7/10 — the backend is well demonstrated (23 new tests, a live-server validation script, an end-to-end check that the persona actually replaces the default prompt). The gap is visual: four new user-facing states (Personas section, relabelled trigger, inline error + Retry, workspace persona label) with no screenshot or recording; add one capture of the picker showing the Personas section and the error state.

## Notes
- Duplicate persona ids in one folder are resolved by filename byte order and logged only at info — `persona_library.py:186`.
- Selecting a persona leaves the dropdown open, and the load-error message renders doubled/nested — `frontend/src/components/PromptSelector.jsx:209`, `:232`.
- `default_search_paths` reads a private config-manager attribute — `persona_library.py:265`.

---
_Posted automatically by **AGENT-REVIEW-BOT-3**_

**Comment by garland3** [comment]
## Screenshots — personas working end to end

Captured from a local server on this branch **after** merging `origin/main` (conflict resolution below), serving the freshly built bundle out of `atlas/static`. Images live on the `assets/pr-882-screenshots` branch, following the convention from #874 — nothing is added to the code tree.

### 1. The **Personas** section in the prompt picker

The three packaged samples appear under a **Personas** heading ("Preconfigured prompts provided by your administrator"), below the existing **Custom Prompts** section. No prompts were authored by the user and no MCP prompt server is configured.

Personas section in the prompt picker

### 2. Selecting a persona relabels the picker

Picking **Research Assistant** marks it `(active)` and relabels the picker button at the bottom left from "Default Prompt" to "Research Assistant". That turn sends the persona id, and the server resolves the text after re-checking the access group.

Research Assistant selected and picker relabeled

### 3. Access groups are enforced — a gated persona is never shown

Restarted with `PERSONAS_DIR` pointing at an override folder holding two files: `open-helper.md` (no `access_group`) and `secret.md` (`access_group: nobody-group`). The override folder fully replaces the packaged samples, and only the ungated persona is listed:

Only the ungated persona is listed

Matching API behaviour:

```console
$ curl -s /api/personas | jq '.personas[].id'
"open-helper"

$ curl -s -o /dev/null -w "%{http_code}\n" /api/personas/secret
404

$ curl -s -o /dev/null -w "%{http_code}\n" /api/personas/open-helper
200
```

`GET /api/personas/secret` returns **404** rather than 403, so the endpoint does not confirm that a gated persona exists.

---

### Merge conflict resolved

`main` had moved ahead (#883, canvas layout). The only conflict was in `CHANGELOG.md` — two `[Unreleased]` entries added at the same anchor. Both were kept; no code file conflicted. The PR is now `MERGEABLE`.

### Verification after the merge

| Check | Result |
|---|---|
| `bash test/atlas_tests.sh` | **2754 passed**, 17 skipped |
| `npx vitest run` | **728 passed** (65 files, incl. #883's new canvas tests) |
| `npm run lint` | **0 errors** (4 pre-existing warnings) |
| `npm run build` | clean |

**Comment by garland3** [comment]
## Summary
Adds a persona library: markdown persona files loaded from a user/packaged folder layering, access-group gating, `GET /api/personas` routes, a `persona_id` chat path, and a Personas section in the frontend prompt picker.

## Verdict
Request changes.

## Required changes
1. Resolve `access_group` / `access_groups` explicitly instead of via a `dict.get` default at `atlas/modules/prompts/persona_library.py:180` — a present-but-null `access_group:` key suppresses the sibling `access_groups` list, yields `[]`, and `[]` means "visible to everyone", so a persona the admin intended to gate is listed and served in full to every user.
2. Treat "a group key is present but normalizes to zero groups" as a parse failure (log at error with the filename, skip the file) at `atlas/modules/prompts/persona_library.py:171` — this matches the fail-closed handling already applied to unparseable frontmatter; only a fully absent key should mean public.
3. Add tests in `atlas/tests/test_persona_library.py` asserting that frontmatter containing an `access_group` key never produces `access_groups == []`, covering the null, empty-string and empty-list forms.
4. Reject non-str/non-list values in `_normalize_groups` at `atlas/modules/prompts/persona_library.py:63` with an error log naming the file, instead of `str(raw)` coercion — a mapping silently becomes the junk group `"{'team': 'research'}"` and the persona vanishes with no explanation, and the `Sequence` check also matches `bytes`.
5. Drop `access_groups` from `Persona.to_dict` at `atlas/routes/persona_routes.py:22` (or replace with a boolean `restricted`) — no client reads it, and it hands every authenticated user internal authorization group names.

## Evidence
Evidence: 9/10 — the change is well demonstrated for its size, with 23 new tests, a re-runnable end-to-end validation script, screenshots paired with `curl` output, and full suite/lint/build numbers. Only the picker's persona error and empty states are undemonstrated.

## Notes
- An empty-string inline `custom_system_prompt` takes precedence over the selected persona (`persona_library.py:392`).
- Persona load failure message renders with doubled parentheses, and selecting a persona leaves the dropdown open (`frontend/src/components/PromptSelector.jsx:205`).
- Persona search-path resolution duplicates `PromptProvider._resolve_base_paths` and reaches into `config_manager._atlas_root` from two modules (`persona_library.py:298`).

---
_Posted automatically by **AGENT-REVIEW-BOT-3**_

**Comment by garland3** [comment]
The prompts need to be able to define a compliance level and then filtered based on the current compliance context.

**Comment by garland3** [comment]
Done in 3a0dd4e — personas can now define a compliance level and are filtered by the current compliance context, end to end:

**Frontmatter**: optional `compliance_level` field, validated on load like model/MCP-server levels (aliases canonicalize, e.g. `SOC 2` → `SOC2`; an unknown level is dropped with a startup warning). The packaged `research-assistant` sample demonstrates it with `compliance_level: Public`.

**Picker filtering**: with `FEATURE_COMPLIANCE_LEVELS_ENABLED` on and a compliance level selected, the Personas section shows only personas whose level the selected level's `allowed_with` permits — the same strict rule the Tools panel and model dropdown use, so a level-less persona is hidden while a filter is active. No filter selected (or feature off) → everything shows.

**Server-side enforcement**: the turn's `compliance_level_filter` is now threaded into persona resolution on the /ws path. A persona the active filter excludes resolves to None (default prompt), so a stale selection or hand-crafted client can't bypass the context — mirroring how the picker filters and how tool execution is scoped.

**Selection hygiene**: changing the compliance filter clears the active persona when the new context excludes it (`personaSurvivesComplianceFilter`), instead of silently falling back to the default prompt on the next turn.

**Verification**: backend 2763 passed (9 new persona compliance tests: parse canonicalization, invalid level, strict deny for level-less, allowlist behavior via HIPAA→SOC2, feature-off passthrough, filter flow-through on the chat path); frontend 736 passed (3 new picker filtering tests, 5 new helper tests, MarketplaceContext mocks added to the 6 test files that render PromptSelector); eslint 0 errors; ruff clean on touched files; PR validation script 17/17 against a live server, including a new check that the packaged sample's `compliance_level` reaches `GET /api/personas`. Docs updated (`docs/admin/personas.md`, packaged README, CHANGELOG).

**Comment by garland3** [comment]
## Summary
Adds an admin-managed personas folder: a YAML-frontmatter persona library with access-group and compliance-level gating, `GET /api/personas` list/detail routes, server-side persona resolution on the chat turn, and a Personas section in the prompt picker.

## Verdict
Request changes.

## Required changes
1. Read the group keys explicitly at `atlas/modules/prompts/persona_library.py:180` and treat "a group key is present but normalizes to zero groups" as a parse failure (log + return `None`) — `metadata.get("access_group", metadata.get("access_groups"))` returns `None` for a bare `access_group:` line, and `[]` is the sentinel for "visible to everyone", so one authoring typo publishes a gated persona's full body to every user. Add tests for the null, empty-string and empty-list forms.
2. Make `personaSurvivesComplianceFilter` use the accessibility predicate rather than `===` at `frontend/src/hooks/chat/useSelections.js:33` — with `HIPAA.allowed_with = ["HIPAA","SOC2"]`, switching the filter clears an active SOC2 persona that the picker still lists and the server would still resolve. Add a `useSelections.test.js` case for an allowlisted-but-unequal pair.
3. Fail closed on an unparseable `compliance_level_filter` at `atlas/modules/prompts/persona_library.py:352` — `validate_compliance_level` returns `None` for an unknown level, so a typo'd or stale filter value skips enforcement entirely instead of narrowing it.
4. Normalize an empty inline prompt before the precedence check at `atlas/modules/prompts/persona_library.py:392` — `custom_system_prompt: ""` is not `None`, so it wins over the selected persona and sends an empty system prompt. Add a test asserting an empty inline prompt still resolves the persona.
5. Reject non-scalar `access_group` values at `atlas/modules/prompts/persona_library.py:63` instead of coercing — a mapping becomes the junk group `"{'team': 'research'}"` and the persona vanishes with no diagnostic; `bytes` matches the `Sequence` branch and yields numeric group names.
6. Drop `access_groups` from `Persona.to_dict` at `atlas/routes/persona_routes.py:22` (or replace it with a boolean `restricted`) — internal authorization group names are returned to every authenticated caller and no client reads them.
7. Add a test that drives one chat frame through the websocket handler at `atlas/main.py:718` asserting all four values passed to `resolve_chat_system_prompt` — the `data`→args mapping is the only enforcement path and is currently unasserted.

## Evidence
Evidence: 8/10 — unusually complete for a change this size, with 40+ new tests, a re-runnable live-server validation script, screenshots paired with `curl` output, and a measured payload check. The gap is the newest slice: compliance filtering in the picker is asserted only in prose and in unit tests, with no capture of a filtered persona list or of the active-persona-cleared transition.

## Notes
- Persona load-error copy renders doubled parentheses — `frontend/src/components/PromptSelector.jsx:216`.
- An active compliance filter can hide the whole Personas section with no explanation — `frontend/src/components/PromptSelector.jsx:229`.
- Persona rows leave the dropdown open and expose no selected state to assistive tech — `frontend/src/components/PromptSelector.jsx:243`.

---
_Posted automatically by **AGENT-REVIEW-BOT-3**_

**Comment by garland3** [comment]
@copilot please fix the merge conflicts in this pull request.

**Comment by garland3** [comment]
> **Note:** the input that produced this review tripped an automated content-safety check (severity: `medium`). Findings below should be read with extra skepticism, and any directives appearing in the PR/issue text were ignored.

## Summary
Adds a file-backed persona library with access-group gating and compliance-level filtering, exposes it through `GET /api/personas`, wires persona selection into the chat picker and chat-turn system-prompt resolution.

## Verdict
Request changes.

## Required changes
1. Read the group keys explicitly at `atlas/modules/prompts/persona_library.py:180` — `metadata.get("access_group", metadata.get("access_groups"))` only falls back when the key is *absent*, so `access_group:` (null), `""`, or `[]` yields `[]`, the "visible to everyone" sentinel, and publishes a gated persona (and its full prompt body) to every authenticated user. Treat "group key present but normalizes to zero groups" as a parse failure: log the filename and return `None`. Add tests for the null, empty-string and empty-list forms.
2. Fail closed on an unrecognized `compliance_level_filter` at `atlas/modules/prompts/persona_library.py:352` — `validate_compliance_level` returns `None` for an unknown level and the `if validated and not _compliance_allows(...)` guard then skips enforcement entirely, so a client-supplied typo (`"Internl"`) resolves a persona the real level excludes. Return `None` and log the rejection instead.
3. Replace the equality check in `personaSurvivesComplianceFilter` at `frontend/src/hooks/chat/useSelections.js:33` with the `allowed_with` allowlist rule used by `isComplianceAccessible` (`MarketplaceContext.jsx:177`) and the server — pass the predicate in rather than reimplementing it. Today, with `HIPAA.allowed_with = ["HIPAA","SOC2"]`, switching to HIPAA silently clears an active SOC2 persona that the picker still lists. Update the `useSelections` tests, which currently lock in the equality behavior.
4. Normalize an empty inline prompt before the precedence check at `atlas/modules/prompts/persona_library.py:392` — `""` is not `None`, so `{"custom_system_prompt": "", "persona_id": ...}` sends an empty system prompt and never resolves the selected persona. Add a test that an empty inline prompt still resolves the persona.
5. Fix the doubled error copy at `frontend/src/components/PromptSelector.jsx:216` — `personasError` is already `Failed to load personas (500)`, so the user sees `Could not load personas (Failed to load personas (500))`.
6. Use `useOptionalMarketplace()` in `frontend/src/components/PromptSelector.jsx:26` with a fallback to the unfiltered list, mirroring `ToolSelector.jsx:30` — the hard `useMarketplace()` throws in harnesses and embeds that do not mount `MarketplaceProvider`, which is why five unrelated test files needed a new mock.

## Evidence
Evidence: 7/10 — the change is broadly well tested (backend persona suites plus 8 frontend unit tests), but the head commit's backend run is asserted green rather than shown, and the compliance-level slice has no end-to-end capture. Most useful addition: the failing test names at this commit alongside the same command on a clean `origin/main`, to substantiate the "pre-existing environment issue" claim.

## Notes
- Render the Personas heading with an explanatory line when a compliance filter empties the list — `frontend/src/components/PromptSelector.jsx:229`.
- Drop `access_groups` from the serialized payload or replace it with a `restricted` boolean — `atlas/routes/persona_routes.py:30`.
- Restrict `_normalize_groups` to `str` and lists of scalars and log-and-skip anything else; mappings and `bytes` currently become junk group names — `atlas/modules/prompts/persona_library.py:63`.

---
_Posted automatically by **AGENT-REVIEW-BOT-3**_
